### PR TITLE
Adding Flowables::empty and Flowables:error

### PR DESCRIPTION
**Pull request by nickgg**
Adding some simple helpers for getting a completed or errored flowable (to assist error handling mostly).

### CODE REVIEW

File: experimental/yarpl/include/yarpl/Flowables.h [hunk 1]
```diff
@@ -77,6 +77,24 @@ class Flowables {
         std::forward<OnSubscribe>(function)));
   }
 
+  template <typename T>
+  static Reference<Flowable<T>> empty() {
+    auto lambda = [](Subscriber<T>& subscriber, int64_t) {
+      subscriber.onComplete();
+      return std::make_tuple(static_cast<int64_t>(0), true);
```
Comment (alexmalyshev) [inline experimental/yarpl/include/yarpl/Flowables.h]: I think `int64_t{0}` is better than `static_cast<int64_t>(0)`.  `int64_t{n}` has the benefit of being a compiler error if `n` cannot be guaranteed to fit within the bounds of an `int64_t`.

File: experimental/yarpl/include/yarpl/Flowables.h [hunk 2]
```diff
@@ -77,6 +77,24 @@ class Flowables {
...
+    };
+    return Flowable<T>::create(std::move(lambda));
+  }
+
+  template <typename T>
+  static Reference<Flowable<T>> error(const std::exception_ptr ex) {
+    auto lambda = [ex](Subscriber<T>& subscriber, int64_t) {
+      subscriber.onError(ex);
+      return std::make_tuple(static_cast<int64_t>(0), true);
```
Comment (alexmalyshev) [inline experimental/yarpl/include/yarpl/Flowables.h]: int64_t{0}

File: experimental/yarpl/test/FlowableTest.cpp [hunk 3]
```diff
@@ -22,12 +22,44 @@ class CollectingSubscriber : public Subscriber<T> {
     values_.push_back(next);
   }
 
+  void onComplete() override {
+    Subscriber<T>::onComplete();
+    complete_ = true;
+  }
+
+  virtual void onError(const std::exception_ptr ex) override {
```
Comment (alexmalyshev) [inline experimental/yarpl/test/FlowableTest.cpp]: `virtual` is unnecessary if you have `override`.

File: experimental/yarpl/test/FlowableTest.cpp [hunk 4]
```diff
@@ -22,12 +22,44 @@ class CollectingSubscriber : public Subscriber<T> {
...
+    Subscriber<T>::onError(ex);
+    error_ = true;
+
+    try {
+      std::rethrow_exception(ex);
+    }
+    catch (const std::exception &e) {
```
Comment (alexmalyshev) [inline experimental/yarpl/test/FlowableTest.cpp]: Nit: `} catch (const std::exception& e) {` is our preferred style.

File: experimental/yarpl/test/FlowableTest.cpp [hunk 5]
```diff
@@ -22,12 +22,44 @@ class CollectingSubscriber : public Subscriber<T> {
...
 
+  bool complete() const {
+    return complete_;
+  }
+
+  bool error() const {
+    return error_;
+  }
+
+  const std::string errorMsg() {
```
Comment (alexmalyshev) [inline experimental/yarpl/test/FlowableTest.cpp]: I believe you want `std::string errorMsg() const {`

File: experimental/yarpl/include/yarpl/Flowables.h [hunk 6]
```diff
@@ -77,6 +77,24 @@ class Flowables {
...
+  static Reference<Flowable<T>> empty() {
+    auto lambda = [](Subscriber<T>& subscriber, int64_t) {
+      subscriber.onComplete();
+      return std::make_tuple(static_cast<int64_t>(0), true);
+    };
+    return Flowable<T>::create(std::move(lambda));
+  }
+
+  template <typename T>
+  static Reference<Flowable<T>> error(const std::exception_ptr ex) {
```
Comment (alexmalyshev) [inline experimental/yarpl/include/yarpl/Flowables.h]: I wonder if you'll also want an overload `Flowables::error(std::exception)`.  Might be a nice wrapper for code like:

`Flowables::error<int>(std::make_exception_ptr(std::runtime_error("oops")))`

File: experimental/yarpl/include/yarpl/Flowables.h (hunk 1 above)
Comment (nickgg) [inline experimental/yarpl/include/yarpl/Flowables.h]: You're worried about the case where zero doesn't fit in an int?

Besides, this matches the existing style within this file.

File: experimental/yarpl/include/yarpl/Flowables.h (hunk 6 above)
Comment (nickgg) [inline experimental/yarpl/include/yarpl/Flowables.h]: Yes I agree, was wondering this when I wrote it but figured there was some reason the interface was changed to exception_ptr.

File: experimental/yarpl/test/FlowableTest.cpp (hunk 3 above)
Comment (nickgg) [inline experimental/yarpl/test/FlowableTest.cpp]: This matches the existing style within this file.

File: experimental/yarpl/test/FlowableTest.cpp (hunk 5 above)
Comment (nickgg) [inline experimental/yarpl/test/FlowableTest.cpp]: sure

File: experimental/yarpl/test/FlowableTest.cpp (hunk 3 above)
Comment (alexmalyshev) [inline experimental/yarpl/test/FlowableTest.cpp]: I should go fix that...

File: experimental/yarpl/include/yarpl/Flowables.h (hunk 1 above)
Comment (alexmalyshev) [inline experimental/yarpl/include/yarpl/Flowables.h]: Okay I should really read this file...

File: experimental/yarpl/include/yarpl/Flowables.h (hunk 6 above)
Comment (alexmalyshev) [inline experimental/yarpl/include/yarpl/Flowables.h]: It's to match `yarpl::flowable::Subscriber::onError(std::exception_ptr)`.  That said, having a second overload for `Flowables::error()` might be convenient.

**Comment by benjchristensen** [comment]
Good additions. Definitely also reminds me we need to create a generic TestSubscriber soon.

**Pull request merged**